### ISSUE TITLE
Server side rendering with universal

### ISSUE THREAD

**Issue by pkozlowski-opensource** [ng-bootstrap/ng-bootstrap#858]
We should explore server-side rendering options with universal. IMO we should be in a pretty good shape to support it as we are not doing anything too crazy. There is direct DOM access in few places but it shouldn't be too hard to remove it.

Anyway, after we release a version with full AoT support (already in master) we should check what it means to support universal. Probably a good starting point would be to use the https://github.com/angular/universal-starter

---

Widgets confirmed to work with Angular 6+:
- [x] Accordion
- [x] Alert
- [x] Buttons
- [x] Carousel (fix in #2494)
- [x] Collapse
- [x] Datepicker*
- [x] Dropdown
- [x] Modal*
- [x] Pagination
- [x] Popover*
- [x] Progressbar
- [x] Rating
- [x] Tabs
- [x] Timepicker
- [x] Tooltip*
- [x] Typeahead*

**Comment by pkozlowski-opensource** [comment]
First report from the battlefield - most things work out of the box
<img width="1433" alt="ng-bootstrap-universal" src="https://cloud.githubusercontent.com/assets/973550/19248832/85aa18ae-8f32-11e6-9095-2763d3a0d3d6.png">
 :-)

**Comment by pkozlowski-opensource** [comment]
So, after trying out all the widgets here are the ones that are not working out of the box:
* [ ] carousel (suspect that one is due to the usage of `setInterval`). In this particular case a page generated by universal never loads (pending / infinitive loop)
* [ ] progressbar - as soon as you add https://github.com/ng-bootstrap/ng-bootstrap/blob/master/src/progressbar/progressbar.ts to the page the whole universal things disappear - not sure what is going on here since this is a very simple directive.
* [ ] typeahead - suspect `Observable.fromEvent` to be the culprit as in this case we are not registering to events through renderer... to be investigated.

All in all the initial results are _very_ encouraging! Didn't expect to get this far without any changes on the ng-bootstrap side (especially that widgets doing _some_ dome manipulation like  modal, tooltip and popovers work great as-is).

**Comment by pkozlowski-opensource** [comment]
Update - actually, typeahead works correctly, it was an error on my part :-) Which means that in the end there are only 2 directives that are troublemakers: 
* carousel - the root cause is well identified and easy to fix
* progressbar - this one needs investigation

**Comment by pkozlowski-opensource** [comment]
Ha! One more update: progressbar works as well, another user error :-) 

Conclusion => fixing carousel will make the entire library universal-compatible 😃

**Comment by MarkPieszak** [comment]
Awesome!! @pkozlowski-opensource 
As Steve mentioned, worst case scenario (for now) you can just wrap that setInterval around isBrowser until we can come up with a better solution for that.

`import { isBrowser } from 'angular2-universal';`

```
if (isBrowser) {
    setInterval(...)
}
```

**Comment by tinchou** [comment]
It shouldn't be too difficult to use
https://github.com/Reactive-Extensions/RxJS/blob/master/doc/api/core/operators/timer.md
 instead!

**Comment by MarkPieszak** [comment]
Could try that, at the same time, don't really need the carousel to "run" on the server, just to paint the controls etc and actually start running on the browser side. Not sure what's the best way to handle all that hmm

**Comment by tinchou** [comment]
Maybe we're just missing some required polyfills:
https://github.com/angular/universal/issues/267

**Comment by JohnGalt1717** [comment]
ngbRadio is still broken with Server side.

**Comment by JohnGalt1717** [comment]
Any word on getting these fixed up? Carousel is still broken too.

**Comment by villanuevadani** [comment]
Is ng-bootstrap pictures slider not working with Universal neither?

**Comment by esistgut** [comment]
I tried the [angular-cli guide](https://github.com/angular/angular-cli/wiki/stories-universal-rendering) to enable universal on an Angular app using ng-bootstrap, I get the following error:

```
$ node server.js 
/myapp/node_modules/@ng-bootstrap/ng-bootstrap/alert/alert.js:1
(function (exports, require, module, __filename, __dirname) { import { Component, Input, Output, EventEmitter, ChangeDetectionStrategy, } from '@angular/core';
                                                              ^^^^^^

SyntaxError: Unexpected token import
    at createScript (vm.js:74:10)
    at Object.runInThisContext (vm.js:116:10)
    at Module._compile (module.js:533:28)
    at Object.Module._extensions..js (module.js:580:10)
    at Module.load (module.js:503:32)
    at tryModuleLoad (module.js:466:12)
    at Function.Module._load (module.js:458:3)
    at Module.require (module.js:513:17)
    at require (internal/module.js:11:18)
    at Object.2RxQ (/myapp/dist-server/main.bf83d26d1ade1e6a9a5a.bundle.js:1:65363)
```

I can work around the problem using webpack to whitelist ng-boostrap with an additional compilation step but I guess this is not the recommended practice.
Note: this is on @angular/cli@1.3.0-rc.3 and @ng-bootstrap/ng-bootstrap@1.0.0-alpha.29.

**Comment by itsrts** [comment]
@esistgut I am also facing the same issue, can you tell if you were successful in making ng-bootstrap whitelist 
I really need this to mke my project move forward

**Comment by esistgut** [comment]
@itsrts https://github.com/angular/angular-cli/issues/1050#issuecomment-321797364

**Comment by dimitriy-k** [comment]
check here some information how the code should be packaged

https://github.com/angular/angular-cli/issues/7200#issuecomment-328991769

**Comment by asgerjensen** [comment]
Disclaimer: I find the whole packaging thing to be a bit of magic, and I may be wrong, but after playing around trying to make this work, I have found three main findings:

- Angulars AOT compiler expects FESM2015 modules. Something I dont know if Webpack2 can create, or even what that means, (aside from the obvious).

- The current module bundles does not seem to be used at runtime, because the current package has the individial .js files directly in the package root dir. That seems to make nodejs read the source files, instead of the package.json when resolving imports/requires.

- I found an old issue on this repo (#542), where a comparison between webpack and rollup was done, and while rollup had the lowest bundle size, webpack was chosen as the package tool. Other literature seems to indicate, that webpack should be the choice for applications, while rollup should be used for libraries, so maybe it would be worth revisiting the rollup vs webpack2 discussion. (I can see that #542 is regarding the demo site, but the main gulp.js uses webpack to package the library also)

I have no problem providing a PR for this, if there is any chance at all it will be accepted, but I would like to have some kind of indication of this, before spending time on it.

**Comment by asgerjensen** [comment]
Maybe, worth mentioning, the main usecase for getting this solved for me, is, that without it, we have to run our SSR version of the site as JIT code. This, in itself is not that bad, since it caches the compiled assets. But it makes it impossible to use i18n with angular/cli, as they only support i18n for AOT compiled code, and I really dont want my main application to know if its running in JIT or AOT mode, and manually handle loading and providing the translations.

**Comment by AlecEJones** [comment]
Hi folks,

The modal service is currently breaking universal builds when injected into components. The issue can be reproduced by cloning the universal-starter repo and injecting the service into any component.

I have written more detail about it in #1968, which has been closed as a duplicate.

Thanks for all your hard work!

**Comment by ilasw** [comment]
@AlecEJones i solved the issue by removing the function "_componentFactoryResolver.resolveComponentFactory" in "node_modules\@ng-bootstrap\ng-bootstrap\modal\modal-stack.js".

This function just dont return the component so change:
` function NgbModalStack(_applicationRef, _injector, _componentFactoryResolver) {
    this._applicationRef = _applicationRef;
    this._injector = _injector;
    this._componentFactoryResolver = _componentFactoryResolver;
    this._windowFactory = this._componentFactoryResolver.resolveComponentFactory(NgbModalWindow); 
    this._backdropFactory = this._componentFactoryResolver.resolveComponentFactory(NgbModalBackdrop);
  }`

Into: 
`function NgbModalStack(_applicationRef, _injector, _componentFactoryResolver) {
		this._applicationRef = _applicationRef;
		this._injector = _injector;
		this._componentFactoryResolver = _componentFactoryResolver;
		this._backdropFactory = (NgbModalBackdrop);
		this._windowFactory = (NgbModalWindow);
	}`

This is not a solution. Just a workaround until the next update.
Have a nice day ;)

**Comment by elSuperRiton** [comment]
@ilasw 
Can you confirm that this fixed the issue ? Did you take any aditionnal steps ?
I've just updated the project from Universal with ang4 to ang5 and now i'm getting : 
" No component factory found for NgbModalBackdrop. Did you add it to @NgModule.entryComponents?  " on the prod build.

**Comment by csentis** [comment]
Same here with `ng-bootstrap-beta.7` and

```shell
Angular CLI: 1.6.0
Node: 8.9.1
OS: win32 x64
Angular: 5.1.0
... animations, common, compiler, compiler-cli, core, forms
... http, language-service, platform-browser
... platform-browser-dynamic, platform-server, router
... service-worker

@angular/cli: 1.6.0
@angular-devkit/build-optimizer: 0.0.35
@angular-devkit/core: 0.0.22
@angular-devkit/schematics: 0.0.41
@ngtools/json-schema: 1.1.0
@ngtools/webpack: 1.9.0
@schematics/angular: 0.1.10
@schematics/schematics: 0.0.10
typescript: 2.5.3
webpack: 3.10.0
```

"No component factory found for NgbModalBackdrop. Did you add it to @NgModule.entryComponents?"

In my `core.module.ts`, I have added the LoginComponent which uses a modal to `entryComponent's` array:

```JS
entryComponents: [
    LoginComponent  // needed by NgbModal
],
```

**Comment by csentis** [comment]
@pkozlowski-opensource I have set up a minimal reproduction repo for the error

```node
No component factory found for NgbModalBackdrop. Did you add it to @NgModule.entryComponents?
```
here: [csentis/ngbmodalbackdrop-universal-issue](https://github.com/csentis/ngbmodalbackdrop-universal-issue).

Repro steps can be found in [the repo's README](https://github.com/csentis/ngbmodalbackdrop-universal-issue/blob/master/README.md).

Really looking forward to you helping here...

**Comment by nishesj** [comment]
One of the workaround @elSuperRiton is to not inject NgbModal in constructor and initialise the service through injector on ngOnInit

`
  if (isPlatformBrowser(this.platformId)) {
      this.modalService = this.injector.get(NgbModal);
    }
`

**Comment by elSuperRiton** [comment]
@nishesj Neat ! How could I miss that. I'll give it a shot and come back to you as soon as it's done. Thanks a lot ;)

**Comment by MatthiasMiebach-Aracom** [comment]
We are currently facing problems when using the tooltip with SSR enabled. 
Every component with a tooltip shows this error when rendered on server side:
ERROR { Error: No component factory found for NgbTooltipWindow. Did you add it to @NgModule.entryComponents?

For Modal the workaround @nishesj mentioned did work 👍 
Any hints or workarounds how we can just disable tooltip on server side?

**Comment by csentis** [comment]
@nishesj Thanks for the idea!

In case anyone wonders how to put [_nishesj_'s line](https://github.com/ng-bootstrap/ng-bootstrap/issues/858#issuecomment-350758405) to work:

```TS
import { PLATFORM_ID } from '@angular/core';
import { isPlatformBrowser } from '@angular/common';

...
// after...
export class CallingModalComponent implements OnInit {
// ... declare modal e.g. with
private modal: any;

// and finally, in constructor
constructor(
        @Inject(PLATFORM_ID) private platformId: Object,
// remember to remove the original reference to NgbModal reference
// private modal: NgbModal // workaround until ng-bootstrap issue 858 gets solved
```

**Comment by cesarsalesgomes** [comment]
@csentis Mixing your code with @nishesj code, i missed the instance for 'injector'.

this.injector.get(NgbModal);

Where i can get?

**Comment by joelnotified** [comment]
@MatthiasMiebach-Aracom Have you found a workaround? I'm also facing the same problem with tooltip not working with AOT for the SSR app

**Comment by joelnotified** [comment]
For anyone facing the same issue as me, I worked around it using this https://github.com/SebastianM/angular-google-maps/issues/1052#issuecomment-331150772

**Comment by CarlosTorrecillas** [comment]
@MatthiasMiebach-Aracom got exaclty the same.  I also implemented the workaround for the modal but indeed it would be nice to get something out for the tooltip. At the moment we cannot render anything with the ngbTooltip in it. The only alternative that I can think of (I haven't tried nor I don't know if that could work) is to use the standard bootstrap tooltip for the time being: https://getbootstrap.com/docs/4.0/components/tooltips/

**Comment by alexkushnarov** [comment]
Angular 5.1.1 with universal server rendering.
I've faced such issue with popover

> ERROR { Error: No component factory found for NgbPopoverWindow. Did you add it to @NgModule.entryComponents?

**How I've fixed that**

I've created `popover.mock.js` file in the `src/server-mocks` dir
mocked file - https://github.com/alexkushnarov/universal-server-mock/blob/master/popover.mock.js

then in the `webpack.config.js` where I configure my server, I've imported NormalModuleReplacementPlugin 
and mocked popover file in the plugins

https://github.com/alexkushnarov/universal-server-mock/blob/master/webpack.config.js

Maybe this will help somebody...
P.S. you can do the same operation for the tooltip and other components where you have errors while server rendering.

**Comment by Ks89** [comment]
I have the same problem with this message:
```
Unexpected token import
/Users/ks89/git/angular-cli-skeleton/node_modules/@ng-bootstrap/ng-bootstrap/accordion/accordion.js:1
(function (exports, require, module, __filename, __dirname) { import { Component, ContentChild, ContentChildren, Directive, EventEmitter, Input, Output, QueryList, TemplateRef } from '@angular/core';
                                                              ^^^^^^

SyntaxError: Unexpected token import
    at createScript (vm.js:80:10)
    at Object.runInThisContext (vm.js:152:10)
    at Module._compile (module.js:624:28)
    at Object.Module._extensions..js (module.js:671:10)
    at Module.load (module.js:573:32)
    at tryModuleLoad (module.js:513:12)
    at Function.Module._load (module.js:505:3)
    at Module.require (module.js:604:17)
    at require (internal/module.js:11:18)
    at Object.PPLX (/Users/ks89/git/angular-cli-skeleton/dist/server/main.077e7be1c26dab452e36.bundle.js:1:87765)
    at n (/Users/ks89/git/angular-cli-skeleton/dist/server/main.077e7be1c26dab452e36.bundle.js:1:211)
    at Object.wnWB (/Users/ks89/git/angular-cli-skeleton/dist/server/main.077e7be1c26dab452e36.bundle.js:1:159255)
    at n (/Users/ks89/git/angular-cli-skeleton/dist/server/main.077e7be1c26dab452e36.bundle.js:1:211)
    at Object.JwoV (/Users/ks89/git/angular-cli-skeleton/dist/server/main.077e7be1c26dab452e36.bundle.js:1:47210)
    at n (/Users/ks89/git/angular-cli-skeleton/dist/server/main.077e7be1c26dab452e36.bundle.js:1:211)
    at Object.Zq8w (/Users/ks89/git/angular-cli-skeleton/dist/server/main.077e7be1c26dab452e36.bundle.js:1:103884)
```

**Comment by slyedoc** [comment]
Is anyone working on this, a dev branch maybe so I am not starting from scratch?

Been working on other stuff waiting for this, but it's been a month and I need SSR before I can go live.

**Comment by AhsanShafique** [comment]
Having issues in ng-bootstrap while implementing https://medium.com/@cyrilletuzi/angular-server-side-rendering-in-node-with-express-universal-engine-dce21933ddce 

D:\ng-bootstrap\ng-bootstrap-Backup\ng-bootstrap>node server.js
D:\ng-bootstrap\ng-bootstrap-Backup\ng-bootstrap\node_modules\@ng-bootstrap\ng-bootstrap\collapse\collapse.js:1
(function (exports, require, module, __filename, __dirname) { import { Directive, Input } from '@angular/core';
                                                              ^^^^^^

SyntaxError: Unexpected token import
    at createScript (vm.js:80:10)
    at Object.runInThisContext (vm.js:139:10)
    at Module._compile (module.js:607:28)
    at Object.Module._extensions..js (module.js:654:10)
    at Module.load (module.js:556:32)
    at tryModuleLoad (module.js:499:12)
    at Function.Module._load (module.js:491:3)
    at Module.require (module.js:587:17)
    at require (internal/module.js:11:18)
    at Object.8Stx (D:\ng-bootstrap\ng-bootstrap-Backup\ng-bootstrap\dist-server\main.bundle.js:1:49653)

**Comment by pripatel93** [comment]
**Solved using work around** (that removes the `SyntaxError: Unexpected token import` error)


Add this to your scripts in `package.json`
```
"transpile:bootstrap": "node node_modules/babel-cli/bin/babel.js node_modules/@ng-bootstrap/ng-bootstrap --out-dir node_modules/@ng-bootstrap/ng-bootstrap --presets es2015"
```
P.S ensure you have babel-cli & babel-preset-es2015 installed

**Comment by chaker89** [comment]
@pripatel93  Hey i am having the same issue, and i tried running your script after installing babel-clu and babel-preset-es2015  All you had to do is run that separately ? then retrigger the build to compile and serve the application ?

**Comment by kvincent3** [comment]
Someone knows how to solve this issue ?

**Comment by Ks89** [comment]
Please could you share something like a schedule for this issue?

Are you waiting for a powerful implementation of angular universal before to update ng-bootstrap or are you working based on the current version of angular universal with all their limitations?

**Comment by WUKS87** [comment]
Please let us know about SSR work.
I have this error:

`
/node_modules/@ng-bootstrap/ng-bootstrap/tooltip/tooltip.js:1
(function (exports, require, module, __filename, __dirname) { import { Component, Directive, Input, Output, EventEmitter, ChangeDetectionStrategy, Injector, Renderer2, ElementRef, ViewContainerRef, ComponentFactoryResolver, NgZone } from '@angular/core';
                                                                     ^

SyntaxError: Unexpected token {
    at new Script (vm.js:74:7)
    at createScript (vm.js:246:10)
    at Object.runInThisContext (vm.js:298:10)
    at Module._compile (internal/modules/cjs/loader.js:646:28)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:689:10)
    at Module.load (internal/modules/cjs/loader.js:589:32)
    at tryModuleLoad (internal/modules/cjs/loader.js:528:12)
    at Function.Module._load (internal/modules/cjs/loader.js:520:3)
    at Module.require (internal/modules/cjs/loader.js:626:17)
    at require (internal/modules/cjs/helpers.js:20:18)
`

Thanks in advance

**Comment by zolakt** [comment]
Are there any updates on this? This is breaking my app completely, and preventing me from upgrading to Angular 6. I somehow got it working with Angular 5 and webpack, but on 6 it's breaking all over the place:

With the CLI, I get the "Unexpected token import" error.
With custom Webpack config (working with Angular5) I get "Cannot read property 'forRoot' of undefined".

This is a huge issue and I'll have to resort to another lib it's don't resolved soon. This turned out to be a very buggy lib for anything related to SSR or AOT. Do you have some ETA at least?

**Comment by LucianoDiPierno** [comment]
Same problem as @zolakt after migration from Angular 5 to 6 and webpack 3 to 4:

> prerendering failed because of error: ReferenceError: Document is not defined

Has anyone found how to solve? Thanks for the help.

**Comment by kvincent3** [comment]
My solution for this issue was to remove ng-bootstrap and use angular material which is fully compatible with angular universal.

**Comment by zolakt** [comment]
@kvincent3 That isn't really a solution is it? It's a completely different framework, not nearly interchangeable. And I'll try to skip the debate how is it better then a framework with 10x more features, because I just can't understand people who claim so

@LucianoDiPierno If you have an ejected webpack config, add the DefinePlugin workaround I posted. It should work

**Comment by kvincent3** [comment]
@zolakt As you said it's not a solution because it's not interchangeable. But I spent days trying to resolve all issues throw by ng-bootstrap and at the end of the day I just give up because it's not compatible. Actually, it took me three days to remove ng-bootstrap and use angular material and I was like you (my app was finished). But I think this is the price to pay when you install ssr at the end of the project. And by the way ng-bootstrap was not the only thing which stopped working after ssr. There are a few number of lib compatible with it. But angular material works perfectly, that's why I let my comment because I was really  surprise.

**Comment by zolakt** [comment]
@kvincent3 Depends on the app really, but I wouldn't bet on your 3 days for most apps. It's not a problem to replace components and JS. When it comes to CSS, Material is not in the same league with Bootstrap. If you used the default Bootstrap theme, and overriden everything manually, maybe you could change it quickly. But if you compiled Bootstrap from SCSS with custom variables, all of your styling goes down the drain. On top of that, add the lack of: responsiveness, utility classes, carousel... and you are styling from scratch all over again

**Comment by LucianoDiPierno** [comment]
@zolakt thanks for the workaround, it work fine

**Comment by dmostert** [comment]
The problem is that the document parameter on NgbFocusTrapFactory constructor is of type Document i.s.o any - Document does not exist in the node process

**Comment by Gajendrasinh** [comment]
I am getting issue with model popup.

anyone fixed this.

![capture](https://user-images.githubusercontent.com/10431380/42133760-c4174f26-7d4c-11e8-80b4-e569ffc3112a.PNG)

**Comment by maxokorokov** [comment]
Added a list of confirmed widgets to the issue, will go through them to make sure they work

**Comment by CarlosTorrecillas** [comment]
@maxokorokov is v3.0 meant to fix this outstanding issue? If not, do you guys have plans to address this in the short term?

**Comment by maxokorokov** [comment]
So, I went through all components and tried pre-rendering all examples we have on the demo site (ex. https://ng-bootstrap.github.io/#/components/accordion/examples), using angular & cli 6.0.8, the universal setup from angular.io and the ng-bootstrap code from master.

I haven't found any major issues with any of components in any of the examples, apart from the fix for the carousel in #2494 that will land in `3.0.0`. Also for custom build setups, note that library will be built and delivered using angular cli in a slightly different format starting from `3.0.0` (#2335 ). 

So I'm going to close this issue after the carousel fix lands.

Note that components marked with (*) in the issue description could be instantiated dynamically (plus optionally attached to the body). We don't support these use cases currently by default (ex. why would you pre-render opened result list in a typeahead or opened datepicker popup?). We would prefer to treat such issues on a case-by-case basis separately, if necessary.

**Comment by maxokorokov** [comment]
@hlohrenz, we'll gladly fix the issue if there is anything to fix on our side. Feel free to help with it as well. I explained in the previous comment the standard setup I was using for testing and there were no issues with dropdown.

So, could you please open a separate issue with your setup and a minimal reproduction scenario, if you think something is wrong with dropdown? You could also try testing with `3.0.0-beta.1`.

**Comment by pkozlowski-opensource** [comment]
. We've tested CLI setup with `3.0.0-beta.1` and managed to get all the widgets working, see: https://github.com/ng-bootstrap/ng-bootstrap/issues/858#issuecomment-405963409. As such I believe that we can close this generic issue now.

If anyone still faces issue with server-side rendering please create a new issue with a minimal reproduce scenario (remember to test with `3.0.0-beta.1`, though.

**Comment by divya130** [comment]
> Please let us know about SSR work.
> I have this error:
> 
> ```
> /node_modules/@ng-bootstrap/ng-bootstrap/tooltip/tooltip.js:1
> (function (exports, require, module, __filename, __dirname) { import { Component, Directive, Input, Output, EventEmitter, ChangeDetectionStrategy, Injector, Renderer2, ElementRef, ViewContainerRef, ComponentFactoryResolver, NgZone } from '@angular/core';
>                                                                      ^
> 
> SyntaxError: Unexpected token {
>     at new Script (vm.js:74:7)
>     at createScript (vm.js:246:10)
>     at Object.runInThisContext (vm.js:298:10)
>     at Module._compile (internal/modules/cjs/loader.js:646:28)
>     at Object.Module._extensions..js (internal/modules/cjs/loader.js:689:10)
>     at Module.load (internal/modules/cjs/loader.js:589:32)
>     at tryModuleLoad (internal/modules/cjs/loader.js:528:12)
>     at Function.Module._load (internal/modules/cjs/loader.js:520:3)
>     at Module.require (internal/modules/cjs/loader.js:626:17)
>     at require (internal/modules/cjs/helpers.js:20:18)
> ```
> Thanks in advance

hi @WUKS87,  even i have same issue how did you solved this issue??

**Comment by LaysDragon** [comment]
> So, I went through all components and tried pre-rendering all examples we have on the demo site (ex. https://ng-bootstrap.github.io/#/components/accordion/examples), using angular & cli 6.0.8, the universal setup from angular.io and the ng-bootstrap code from master.
> 
> I haven't found any major issues with any of components in any of the examples, apart from the fix for the carousel in #2494 that will land in `3.0.0`. Also for custom build setups, note that library will be built and delivered using angular cli in a slightly different format starting from `3.0.0` (#2335 ).
> 
> So I'm going to close this issue after the carousel fix lands.
> 
> Note that components marked with (*) in the issue description could be instantiated dynamically (plus optionally attached to the body). We don't support these use cases currently by default (ex. why would you pre-render opened result list in a typeahead or opened datepicker popup?). We would prefer to treat such issues on a case-by-case basis separately, if necessary.

For my case, our designer use modal as a new page to show up in front of other page.  With properly wrapper, I made it work with router simply. Then now we bumped into the ssr problem. 🤔

**Comment by maxokorokov** [comment]
Locking this one, if you have any problems, please report it as a separate issue. It will be more productive and won't get lost.
